core/gather/gatherers/trace-elements.js

```diff
@@ -65,6 +65,61 @@ class TraceElements extends BaseGatherer {
     if (name) this.animationIdToName.set(id, name);
   }
 
+  /**
+   * @param {LH.Artifacts.TraceEngineResult} traceEngineResult
+   * @param {string} mainFrameId
+   * @return {Promise<Array<{nodeId: number}>>}
+   */
+  static async getTraceEngineElements(traceEngineResult, mainFrameId) {
+    // Can only resolve elements for the latest insight set, which should correspond
+    // to the current frame id. Can't resolve elements for pages that are gone.
+    const insightSet = [...traceEngineResult.insights.values()]
+      .reverse().find(insightSet => insightSet.frameId === mainFrameId);
+    if (!insightSet) {
+      return [];
+    }
+
+    /**
+     * Execute `cb(obj, key)` on every object property (non-objects only), recursively.
+     * @param {any} obj
+     * @param {(obj: Record<string, string>, key: string) => void} cb
+     * @param {Set<object>} seen
+     */
+    function recursiveObjectEnumerate(obj, cb, seen) {
```

Comment:
I think there is some risk that we will resolve node details for a bunch of nodes we don't necessarily need. An insight might hold on to a trace event with a `nodeId` that never gets used in one of the insights.

I guess the alternative is to manually curate the list of `nodeId`s but then we would potentially miss some. False positives > false negatives so this seems good.